File: 1020-number-of-enclaves/1020-number-of-enclaves.cpp
class Solution {
public:
    void dfs(int r, int c, vector<vector<int>>& grid, vector<vector<int>> &visited){
        
        if(r<0 || r>=grid.size() || c<0 || c>=grid[0].size() || grid[r][c] == 0 || visited[r][c] == 1)
            return;
        
        visited[r][c] = 1;
        
        dfs(r-1, c, grid, visited);
        dfs(r, c-1, grid, visited);
        dfs(r+1, c, grid, visited);
        dfs(r, c+1, grid, visited);
    }
    
    int numEnclaves(vector<vector<int>>& grid) {
        
        int n = grid.size();
        int m = grid[0].size();
        
        vector<vector<int>> visited(n, vector<int> (m, 0));
        // dfs on boundary
        // column boundary
        for(int row=0; row<n; row++){
            if(grid[row][0] == 1 and visited[row][0] == 0)
                dfs(row, 0, grid, visited);
            
            if(grid[row][m-1] == 1 and visited[row][m-1] == 0)
                dfs(row, m-1, grid, visited);
        }
        
        // row boundary
        for(int col=0; col<m; col++){
            if(grid[0][col] == 1 and visited[0][col] == 0)
                dfs(0, col, grid, visited);
            
            if(grid[n-1][col] == 1 and visited[n-1][col] == 0)
                dfs(n-1, col, grid, visited);
        }
        
        int count = 0;
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                
                if(grid[i][j] == 1 and visited[i][j] == 0)
                    count++;
            }
        }
        
        return count;
    }
};

File: 1020-number-of-enclaves/1020-number-of-enclaves.java
class Solution {
    private void bfs(Queue<int[]> q, int[][] grid, int[][] vis, int[][] delta){
        
        int N = grid.length;
        int M = grid[0].length;
        
        while(q.size()>0){
            
            int row = q.peek()[0];
            int col = q.peek()[1];
            q.remove();
            
            for(int i=0; i<4; i++){
                int nrow = row + delta[i][0];
                int ncol = col + delta[i][1];
                
                if(nrow >= 0 && ncol >= 0 && nrow < N && ncol < M && grid[nrow][ncol] == 1 && vis[nrow][ncol] == 0){
                    q.add(new int[] {nrow, ncol});
                    vis[nrow][ncol] = 1;
                }
            }
        }
    }
    
    private void findEnclavesInGrid(int[][] grid, int[][] vis, int[] cnt){
        
        int N = grid.length;
        int M = grid[0].length;
        
        for(int i=0; i<N; i++){
            for(int j=0; j<M; j++){
                if(grid[i][j] == 1 && vis[i][j] == 0){
                    cnt[0]++;
                }
            }
        }
    }
    
    public int numEnclaves(int[][] grid) {
        
        int N = grid.length;
        int M = grid[0].length;
        
        int[][] vis = new int[N][M];
        Queue<int[]> q = new LinkedList<>();
        
        // travel for boundary rows
        for(int i=0; i<N; i++){
            for(int j=0; j<M; j++){
                
                if(i==0 || j==0 || i==N-1 || j==M-1){
                    if(grid[i][j] == 1){
                        q.add(new int[] {i, j});
                        vis[i][j] = 1;
                    }
                }
            }
        }
        
        
        int[][] delta = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
        bfs(q, grid, vis, delta);
        
        int[] cnt = {0};
        findEnclavesInGrid(grid, vis, cnt);
        return cnt[0];
    }
}